qiskit/algorithms/minimum_eigensolvers/vqe.py

```diff
@@ -0,0 +1,380 @@
+
+        start_time = time()
+
+        evaluate_energy = self._get_evaluate_energy(ansatz, operator)
+
+        if self.gradient is not None:
+            evaluate_gradient = self._get_evalute_gradient(ansatz, operator)
+        else:
+            evaluate_gradient = None
+
+        # perform optimization
+        if callable(optimizer):
+            opt_result = optimizer(fun=evaluate_energy, x0=initial_point, jac=evaluate_gradient)
+        else:
+            opt_result = optimizer.minimize(
+                fun=evaluate_energy, x0=initial_point, jac=evaluate_gradient
+            )
+
+        eval_time = time() - start_time
+
+        result = VQEResult()
+        result.eigenvalue = opt_result.fun + 0j
+        result.cost_function_evals = opt_result.nfev
+        result.optimal_point = opt_result.x
+        result.optimal_parameters = dict(zip(ansatz.parameters, opt_result.x))
+        result.optimal_value = opt_result.fun
+        result.optimizer_time = eval_time
+
+        logger.info(
+            "Optimization complete in %s seconds.\nFound opt_params %s",
+            eval_time,
+            result.optimal_point,
+        )
+
+        if aux_operators:
+            # not None and not empty list or dict
+            bound_ansatz = ansatz.bind_parameters(opt_result.x)
+            # aux_values = eval_observables(self.estimator, bound_ansatz, aux_operators)
+            # TODO remove once eval_operators have been ported.
+            aux_values = self._eval_aux_ops(bound_ansatz, aux_operators)
+            result.aux_operator_eigenvalues = aux_values
+
+        return result
+
+    @classmethod
+    def supports_aux_operators(cls) -> bool:
+        return True
+
+    def _get_evaluate_energy(
+        self,
+        ansatz: QuantumCircuit,
+        operator: BaseOperator | PauliSumOp,
+    ) -> tuple[Callable[[np.ndarray], float | list[float]], dict]:
+        """Returns a function handle to evaluate the energy at given parameters for the ansatz.
+        This is the objective function to be passed to the optimizer that is used for evaluation.
+        Args:
+            ansatz: The ansatz preparing the quantum state.
+            operator: The operator whose energy to evaluate.
+        Returns:
+            Energy of the Hamiltonian of each parameter.
+        Raises:
+            AlgorithmError: If the primitive job to evaluate the energy fails.
+        """
+        num_parameters = ansatz.num_parameters
+
+        # avoid creating an instance variable to remain stateless regarding results
+        eval_count = 0
+
+        def evaluate_energy(parameters):
+            nonlocal eval_count
+
+            # handle broadcasting: ensure parameters is of shape [array, array, ...]
+            parameters = np.reshape(parameters, (-1, num_parameters)).tolist()
+            batchsize = len(parameters)
+
+            try:
+                job = self.estimator.run(batchsize * [ansatz], batchsize * [operator], parameters)
```

Comment:
These are the same ansatz and operator run over different (a set of) parameters. At present the ansatz and operator must be replicated in the list (does not support broadcasting). The primitive will manage a cache of circuits it has seen to transpile them just once etc. The circuit is passed by reference and the primitive has to figure if they are the same etc.

Reply:
Okay I see, thanks! I mostly brought this up so that we can ensure that the parameter binding on e.g. the first list item, does not also bind parameters on all the others :smile: